-fix compile_thrift.py

shengofsun opened the pull request:
-smallfix in code generator
-other small bug fix
-compile options from "c++11" to "c++1y" to enable some c++14 features
-resolve #422

imzhenyu commented:
@shengofsun we take quite some time deciding upgrading to C++11, so I'm afraid we have to do that again for C++14. What are the lowest compiler versions that support C++14? Do the features really necessary?

shengofsun committed:
-fix compile_thrift.py
-smallfix in code generator
-other small bug fix
-compile options from "c++11" to "c++1y" to enable some c++14 features

imzhenyu commented on src/core/core/app_manager.cpp:
```diff
@@ -247,7 +243,7 @@ namespace dsn
             }
             else
-                return false;
+                return nullptr;
```
return type is 'bool'.

shengofsun replied:
Sorry I didn't pay much attention on this, but the function prototype is "rpc_handler_info\* app_manager::rpc_unregister_handler(dsn_gpid gpid, dsn_task_code_t rpc_code)", which is not "bool".

imzhenyu commented on bin/dsn.cmake:
```diff
@@ -231,7 +231,7 @@ endmacro(ms_replace_compiler_flags)
     if(UNIX)
         include(CheckCXXCompilerFlag)        
-        CHECK_CXX_COMPILER_FLAG("-std=c++11" COMPILER_SUPPORTS_CXX11)
+        CHECK_CXX_COMPILER_FLAG("-std=c++1y" COMPILER_SUPPORTS_CXX11)
```
let's stick to C++11 to avoid asking developers too much about compilers. Unless you have something that is very important and cannot be implemented using C++11 and below.

shengofsun replied:
Actually we are using C++14 features, in which we can use move semantic in lambda capture list, such as things like this: "[ b = std::move(c)]".
Well, the compile flag "-std=c++11" works fine on my desktop with the g++ version of 4.8. But on my laptop with the g++ version of 5.1, "-std=c++11" won't compile.

imzhenyu replied:
My bad, it is because of move lambda that we abandon VS 2013 and use VS 2015 last time. I thought it was part of the C++11 features when seeing this change as the change should have been done already last time. You have just fixed a bug:)
